api-v2/src/student/student.service.ts

```diff
@@ -0,0 +1,108 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import {
+  DeleteResult,
+  FindOneOptions,
+  Repository,
+  UpdateResult,
+} from 'typeorm';
+import { CreateStudentDto } from './dto/create-student.dto';
+import { UpdateStudentDto } from './dto/update-student.dto';
+import { Student } from './entities/student.entity';
+import { StudentNotFoundError } from './student-not-found.error';
+
+@Injectable()
+export class StudentService {
+  constructor(
+    @InjectRepository(Student)
+    private studentRepository: Repository<Student>,
+  ) {}
+
+  async create(createStudentDto: CreateStudentDto): Promise<Student> {
+    // make sure the user doesn't already exists
+    const { email } = createStudentDto;
+    const userInDb = await this.studentRepository.findOne({ where: { email } });
+    if (userInDb) {
+      throw new Error('A user with the email is already registered');
```

Comment:
we should put error messages into a constant, just so changing them is easier later on. something like ERROR_MESSAGES.controllers.student.alreadyExists. 

also here, why throw Error, instead of HttpError? i saw in earlier places in code, try/catch around calling service methods and then throwing HttpError (assumes service just throws normal Errors), but also some throwing HttpError in (other) service methods (assumes no try/catch around service method call). we should pick one convention, and then stick with it. two good options:

- ignore "controllers are http only layer" and just throw HttpError in services, always–this would then mean no try/catches around service method calls, since we have to propagate the HttpErrors to the controller
- be good citizens and customize return values; save exceptions for exceptional behavior (don't do throwing in services generally, just return null/something to signify what happened), then throw HttpError/handle it in controller. probably more work, but probably cleaner as well.

Reply:
I like both these options, would like to discuss them a little more during GM!

My current implementation has 2 intentions
- Catching issues in a method as soon as possible and throwing an error instead of relying on classes using the method to throw errors
- Not throwing HttpError in service

I may have by mistake thrown an HttpError in a service but i can't seem to find it but that definitely isn't my intention.